File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob4.tar/Prob4/Prob4/Code/c_parallel/multigrid.c
#include <stdio.h>
#include "multigrid.h"

#define URF(I,J) uf[(I)*(2*ny)+(J)]
#define URC(I,J) uc[(I)*(ny)+(J)]
#define UIC(I,J) ucwithghost[(I)*(nc_y+2)+(J)]
#define UIF(I,J) uf[(I)*(ny)+(J)]

void restriction(double *uf, double *uc, int nx, int ny)
{
    int ic, jc;

    for (ic=0; ic<nx; ic++) 
    {
	for (jc=0; jc<ny; jc++) 
	{
	    URC(ic,jc) = 1.0/4.0*( \
		    URF(2*ic,2*jc)+URF(2*ic+1,2*jc) \
		    +URF(2*ic,2*jc+1)+URF(2*ic+1,2*jc+1));
	}
    }

    MPI_Barrier(MPI_COMM_WORLD);
}

void interpolation(double *uc, double *uf, int nx, int ny, int myrank, int ncpus)
{
    int ic, jc, nc_x, nc_y, count=0;
    int leftrank, rightrank;
    double *ucwithghost;
    int i,j;
    MPI_Request req[2];
    MPI_Status status[2];

    nc_x = nx/2;
    nc_y = ny/2;
   
    leftrank = (myrank-1+ncpus)%ncpus;
    rightrank = (myrank+1)%ncpus;

    ucwithghost = (double*)malloc((nc_x+2)*(nc_y+2)*sizeof(double));

    for (i=1; i<nc_x+1; i++) 
    {
	for (j=1; j<nc_y+1; j++) 
	{
	    UIC(i,j) = uc[count];
	    count++;
	}
    }
    
    for (i=1; i<nc_x+1; i++) 
    {
	UIC(i,0) = -UIC(i,1);
	UIC(i,nc_y+1) = -UIC(i,nc_y);
    }

    if(ncpus>1) 
    {
	MPI_Irecv(&UIC(nc_x+1,0), nc_y+2, MPI_DOUBLE, rightrank, 1002, MPI_COMM_WORLD, &req[0]);
	MPI_Isend(&UIC(1,0), nc_y+2, MPI_DOUBLE, leftrank, 1002, MPI_COMM_WORLD, &req[1]);
	MPI_Waitall(2, req, status);
	
        MPI_Irecv(&UIC(0,0), nc_y+2, MPI_DOUBLE, leftrank, 1002, MPI_COMM_WORLD, &req[0]);
        MPI_Isend(&UIC(nc_x,0), nc_y+2, MPI_DOUBLE, rightrank, 1002, MPI_COMM_WORLD, &req[1]);
        MPI_Waitall(2, req, status);	
    }

    for (j=1; j<nc_y+1; j++) 
    {
	if(myrank==0)
		UIC(0,j) = -UIC(1,j);
	if(myrank==ncpus-1)
		UIC(nc_x+1,j) = -UIC(nc_x,j);
    }

    if(myrank==0)
    {
	UIC(0,0) = -UIC(1,1);
    	UIC(0,nc_y+1) = -UIC(1,nc_y);
    }

    if(myrank==ncpus-1)
    {
    	UIC(nc_x+1,0) = -UIC(nc_x,1);
    	UIC(nc_x+1,nc_y+1) = -UIC(nc_x,nc_y);
    }

    for (ic=0; ic<nc_x; ic++) 
    {
	for (jc=0; jc<nc_y; jc++) 
	{
	    UIF(2*ic,2*jc) = (9.0*UIC(ic+1,jc+1) \
		    + 3.0*UIC(ic,jc+1) \
		    + 3.0*UIC(ic+1,jc) \
		    + 1.0*UIC(ic,jc))/16.0;

	    UIF(2*ic+1,2*jc) = (9.0*UIC(ic+1,jc+1) \
		    + 3.0*UIC(ic+2,jc+1) \
		    + 3.0*UIC(ic+1,jc) \
		    + 1.0*UIC(ic+2,jc))/16.0;

	    UIF(2*ic,2*jc+1) = (9.0*UIC(ic+1,jc+1) \
		    + 3.0*UIC(ic,jc+1) \
		    + 3.0*UIC(ic+1,jc+2) \
		    + 1.0*UIC(ic,jc+2))/16.0;

	    UIF(2*ic+1,2*jc+1) = (9.0*UIC(ic+1,jc+1) \
		    + 3.0*UIC(ic+2,jc+1) \
		    + 3.0*UIC(ic+1,jc+2) \
		    + 1.0*UIC(ic+2,jc+2))/16.0;
	}
    }

    free(ucwithghost);

    MPI_Barrier(MPI_COMM_WORLD);
}

File: KSC2016_ProblemSet_n_Solution/Prob4/Code/f_parallel/multigrid.f90
module multigrid

    use mpi

    implicit none
    
contains

    subroutine restriction(uf,uc,nx,ny)
    
    implicit none
    
    real(8) :: uf(2*nx,2*ny), uc(nx,ny)
    integer(4) :: nx,ny
    integer(4) :: ic, jc, ierr
    
    do jc=1, ny
        do ic=1,nx
            uc(ic,jc) = 1.0/4.0*(uf(2*ic-1,2*jc-1)+uf(2*ic,2*jc-1)+uf(2*ic-1,2*jc)+uf(2*ic,2*jc))
        enddo
    enddo

    call MPI_BARRIER(MPI_COMM_WORLD,ierr)
    
    end subroutine restriction
    
    subroutine interpolation(uc,uf,nx,ny,myrank,ncpus)
    
    implicit none

    real(8) :: uf(nx,ny), uc(nx/2,ny/2)
    integer(4) :: nx,ny,myrank,ncpus
    integer(4) :: i,j,ic,jc,ncx,ncy,count=0
    integer(4) :: prevrank, nextrank
    integer(4) :: ierr, request(2), status(MPI_STATUS_SIZE)
    real(8),allocatable :: uc_ghost(:,:)

    ncx = nx/2
    ncy = ny/2

    prevrank = mod(myrank-1+ncpus,ncpus)
    nextrank = mod(myrank+1,ncpus)

    allocate(uc_ghost(0:ncx+1,0:ncy+1))

    do jc=1,ncy
        do ic=1,ncx
            uc_ghost(ic,jc) = uc(ic,jc)
        enddo
    enddo

    do jc=1,ncy
        uc_ghost(0,jc) = -uc_ghost(1,jc)
        uc_ghost(ncx+1,jc) = -uc_ghost(ncx,jc)
    enddo


    if(ncpus.gt.1) then
        call MPI_IRECV(uc_ghost(0,ncy+1),ncx+2,MPI_REAL8,nextrank,1002,MPI_COMM_WORLD,request(1),ierr)
        call MPI_ISEND(uc_ghost(0,1),ncx+2,MPI_REAL8,prevrank,1002,MPI_COMM_WORLD,request(2),ierr)
        call MPI_WAITALL(2,request,status,ierr)

        call MPI_IRECV(uc_ghost(0,0),ncx+2,MPI_REAL8,prevrank,1002,MPI_COMM_WORLD,request(1),ierr)
        call MPI_ISEND(uc_ghost(0,ncy),ncx+2,MPI_REAL8,nextrank,1002,MPI_COMM_WORLD,request(2),ierr)
        call MPI_WAITALL(2,request,status,ierr)
    endif

    do ic=1,ncx
        if(myrank.eq.0) then
            uc_ghost(ic,0) = -uc_ghost(ic,1)
        endif
        if(myrank.eq.ncpus-1) then
            uc_ghost(ic,ncy+1) = -uc_ghost(ic,ncy)
        endif
    enddo

    if(myrank.eq.0) then
        uc_ghost(0,0) = -uc_ghost(1,1);
        uc_ghost(ncx+1,0) = -uc_ghost(ncx,1);
    endif
    if(myrank.eq.ncpus-1) then
        uc_ghost(0,ncy+1) = -uc_ghost(1,ncy);
        uc_ghost(ncx+1,ncy+1) = -uc_ghost(ncx,ncy);
    endif

    do jc=1,ncy
        do ic=1,ncx
            uf(2*ic-1,2*jc-1) = (9.0*uc_ghost(ic,jc) + 3.0*uc_ghost(ic-1,jc) + 3.0*uc_ghost(ic,jc-1) + 1.0*uc_ghost(ic-1,jc-1))/16.0
            uf(2*ic,  2*jc-1) = (9.0*uc_ghost(ic,jc) + 3.0*uc_ghost(ic+1,jc) + 3.0*uc_ghost(ic,jc-1) + 1.0*uc_ghost(ic+1,jc-1))/16.0
            uf(2*ic-1,2*jc)   = (9.0*uc_ghost(ic,jc) + 3.0*uc_ghost(ic-1,jc) + 3.0*uc_ghost(ic,jc+1) + 1.0*uc_ghost(ic-1,jc+1))/16.0
            uf(2*ic,  2*jc)   = (9.0*uc_ghost(ic,jc) + 3.0*uc_ghost(ic+1,jc) + 3.0*uc_ghost(ic,jc+1) + 1.0*uc_ghost(ic+1,jc+1))/16.0
        enddo
    enddo
    
    deallocate(uc_ghost);

    call MPI_BARRIER(MPI_COMM_WORLD,ierr)
    
    end subroutine interpolation
    
end module multigrid
    
